src/ordered.jl

```diff
@@ -8,35 +8,94 @@ When applying the algorithm to an image, the threshold matrix is repeatedly tile
     mat::T
+    color_error_multiplier::R
 end
-function OrderedDither(mat; invert_map=false)
+function OrderedDither(mat; invert_map=false, color_error_multiplier=0.5)
     if invert_map
-        return OrderedDither(1 .- mat)
+        return OrderedDither(1 .- mat, color_error_multiplier)
     end
-    return OrderedDither(mat)
+    return OrderedDither(mat, color_error_multiplier)
 end
 
 function binarydither!(alg::OrderedDither, out::GenericGrayImage, img::GenericGrayImage)
     # eagerly promote to the same eltype to make for-loop faster
     FT = floattype(eltype(img))
     mat = FT.(alg.mat)
 
+    T = eltype(out)
+    black, white = T(0), T(1)
+
+    # Precompute lookup tables for modulo indexing of threshold matrix
     matsize = size(mat)
     rlookup = [mod1(i, matsize[1]) for i in 1:size(img)[1]]
     clookup = [mod1(i, matsize[2]) for i in 1:size(img)[2]]
 
-    T = eltype(out)
-    black, white = T(0), T(1)
-
     @inbounds @simd for i in CartesianIndices(img)
         r, c = Tuple(i)
         out[i] = ifelse(img[i] > mat[rlookup[r], clookup[c]], white, black)
     end
     return out
 end
 
+# Using Pattern Dithering by Thomas Knoll / Adobe Inc. Patent expired in 2019.
+# https://patents.google.com/patent/US6606166B1/en
+# Implemented according to Joel Yliluoma's pseudocode
+# https://bisqwit.iki.fi/story/howto/dither/jy/
+function colordither(
+    alg::OrderedDither,
+    img::GenericImage,
+    cs::AbstractVector{<:Pixel},
+    metric::DifferenceMetric,
+)
+    cs_lab = Lab.(cs)
+    cs_xyz = XYZ.(cs)
+
+    # Precompute lookup tables for modulo indexing of threshold matrix
+    mat = numerator.(alg.mat)
+    nmax = maximum(mat)
+    matsize = size(mat)
+    rlookup = [mod1(i, matsize[1]) for i in 1:size(img)[1]]
+    clookup = [mod1(i, matsize[2]) for i in 1:size(img)[2]]
+
+    # Allocate matrices
+    candidates = Array{UInt8}(undef, nmax)
+    index = Matrix{UInt8}(undef, size(img)...)
+
+    @inbounds for I in CartesianIndices(img)
+        r, c = Tuple(I)
+        err = zero(XYZ)
+        px = XYZ(img[I])
+
+        for j in 1:nmax
+            col = px + alg.color_error_multiplier * err
+            idx = _closest_color_idx(col, cs_lab, metric)
+
+            # We are in loop (idx ↔ err) if we already computed `idx` as the closest color
+            # after the first iteration. Breaking out of this loops lets us avoid calling
+            # the expensive closest color computation.
+            if (j > 2) && (idx in candidates[2:(j - 1)])
+                # Find the range of the loop in `candidates`:
+                looprange = (findall(i -> i == idx, candidates[2:(j - 1)])[1] + 1):(j - 1)
+                # Fill the rest of `candidates` with this loop:
+                lengthfill = nmax - j + 1
+                nrepeat = div(lengthfill, length(looprange), RoundUp)
+                candidates[j:end] = view(
+                    repeat(candidates[looprange]; outer=nrepeat), 1:lengthfill
+                )
```

Comment:
Is there a more efficient way to fill the rest of the array with the loop `candidates[looprange]`?

Reply:
When this is done, I will merge it so we can use the Benchmarking setup for further performance tweaks. 
I will leave our open conversations unresolved for easy referencing in the future as they contain some great ideas. 👍

Reply:
One way is to use a lazy iterator, e.g., `Iterators.repeated`.

If you want, you can also bring a `repeat!` PR to julialang/julia repo; this should be low-hanging fruit and would be a great experience if you haven't done this before.

Reply:
I went with `Iterators.take(Iterators.cycle(loop), n)`, however the performance seems unchanged.